_posts/2022-01-29-stream-processing-part1.md

```diff
@@ -0,0 +1,388 @@
+    ("00:03:00", ("foo", 2L)),
+    ("00:03:00", ("bar", 2L)),
+  ))
+  
+  results.withTimestamp should inOnTimePane("00:01:00", "00:02:00") {
+    beEmpty
+  }
+}
+```
+
+* If there is no input lines for given period, no results are produced
+
+It is quite problematic trait of the streaming pipelines for the frameworks.
+How to recognize if the pipeline is stale from the situation when everything works smoothly but there is no data for some period of time?
+It becomes especially important for joins, if there is no incoming data in one stream it can not block the pipeline forever.
+
+
+## Late Data
+
+I'm glad that you are still there, the most interesting part of this blog post starts here.
+
+Late data is an inevitable part of every streaming application. 
+Imagine that our stream comes from mobile application and someone is on a train that has hit long tunnel somewhere in the Alps ...
+
+Streaming pipeline needs to materialize results in a timely manner, how long the pipeline should wait for data?
+If 99<sup>th</sup> percentile latency is 3 seconds, it does not make any sense to wait for outliers late by minutes or hours. 
+For unbounded data this is always heuristic calculation, 
+the streaming framework continuously estimates time "X" where all input data with event-time less than "X" have been already observed.
+The time "X" is called **watermark**. 
+The watermark calculation algorithm determines quality of the streaming framework. 
+With better heuristics you will get lower overall latency and more precise late data handling. 
+
+Fortunately it's quite easy to write fully deterministic test scenario for late data. 
+Good streaming frameworks (like Apache Beam) provide watermark programmatic control for testing purposes.
+
+```scala
+val DefaultWindowDuration = Duration.standardMinutes(1L)
+
+"Late words" should "be silently dropped" in runWithContext { sc =>
+  val words = testStreamOf[String]
+    .addElementsAt("00:00:00", "foo bar")
+    .addElementsAt("00:00:30", "baz baz")
+    .advanceWatermarkTo("00:01:00")
+    .addElementsAt("00:00:40", "foo") // late event
+    .advanceWatermarkToInfinity()
+  
+  val results = wordCountInFixedWindow(sc.testStream(words), DefaultWindowDuration)
+  
+  results.withTimestamp should containInAnyOrderAtWindowTime(Seq(
+    ("00:01:00", ("foo", 1L)),
+    ("00:01:00", ("bar", 1L)),
+    ("00:01:00", ("baz", 2L)),
+  ))
+}
+```
+
+* After "foo bar" and "baz baz" on-time events, the watermark is programmatically advanced to the end of the first one-minute window
+* As an effect of updated watermark, the results for the first window are materialized.
+* Then late event "foo" is observed, it should be included in the results of window "00:01:00" but it is silently dropped!
+
+With high quality heuristic watermark it should be rare situation that watermark is advanced too early. 
+But as a developer you have to take into account this kind of situation.
+
+## Late Data Under Allowed Lateness (Discarded)
```

Comment:
Gdy przeczytałem po raz pierwszy, uznałem że ten late event zostanie odrzucony.
Może warto tu dodać `(Fired Panes Discarder)`? To komplikuje tytuł, ale ogranicza dwuznaczność.